wip: Rewrite to more modular CLI

aparcar opened the pull request:
This is a rewrite of the CLI to make it more modular and easier to maintain. Initial motivation was to create `man` pages automatically. 

The work started prior the *broker* API stuff, so that part is missing right now - if the general structure is acceptable I'd implement that, too.

Some assumptions:
* Use `clap_mangen` to generate documentation instead of "raw" files.
* Use `clap-verbosity-flag` for better handling log levels. I.e. you add `-q` for quite or `-vv` to be more verbose than `-v`.
* Have `cli.rs` as abstract thing only, don't add any functionality to it.
* Move `event_loop.rs` into its own file (included via `cmd/exchange.rs` and `cmd/exchangeconfig.rs`). Should it be added to `cmd/` instead?
* ~~Dropped deprecated `Keygen` command~~

**I expect pretty much all tests to fail at this stage**

I think with the [derive] approach it's not fully possible to add sections like "BUGS" and "STANDARDS", I may use partly the "builder" approach or create upstream PRs against `clap`.

aparcar committed:
Rewrite to more modular CLI

Signed-off-by: Paul Spooren <mail@aparcar.org>

koraa commented:
Thank you for submitting this PR; it looks very nice.

What worries me a bit is the lack of integration tests checking whether the CLI keeps functioning, so its hard to figure out whether this change could introduce any regression.

> Dropped deprecated Keygen command

Lets keep it for now for backwards compatibility. There is quite a number of deployments at this point so its quite important not to throw away stuff unless its really necessary.

aparcar commented:
@koraa this seems to slowly getting into shape, one thing I'm currently troubled with is the dynamic import of the ApiCli. Using the current approach for `build.rs`, this causes errors. I'm not familiar enough with the build system here so if anyone has an idea, please help out.

aparcar committed:
wip: cleanups

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
chore: fd.rs import bail only on linux OS

Bail is only used on linux targets.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
wip: print help by default, drop build.rs, add man arg

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar commented:
As discussed via chat I dropped `build.rs` and moved it to `main.rs`, which now allows to use `--print-manpage` and `--print-completions`, which was a long hanging fruit anyway.

By using `arg_required_else_help` from `clap`, the help page is printed by default which was a outstanding TODO anyway https://github.com/rosenpass/rosenpass/blob/main/rosenpass/src/cli.rs#L223

I took the freedom to call the `get_brooker_interface` only when actually used, i.e. when `exchange` or `exchange-config` is called.

aparcar commented:
For packaging, this gives a good idea how OS could handle completions and man pages: https://gitlab.archlinux.org/archlinux/packaging/packages/genact/-/blob/main/PKGBUILD?ref_type=heads

aparcar committed:
wip, rfc: drop man command

the command isn't super useful since it just call `man` and whoever
packages the thing should take care of man page generation (via
--print-manpage).

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
wip: fix cargo doc

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
ci: Skip Nix build of aarch64 since it takes forever

More than 6 hours aka failing the CI. Drop it for now and hope to have
it enabled later again.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
ci: Drop mandoc job for rosenpass.1

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
event_loop: pass broker_interface and test_helpers

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
wip: cleanup create_broker()

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
ci: Rename regression workflow to "Regression"

No magic here, this is likely a copy&paste error. Problem is that one
workflow being called "QC" (regressions.yml) cancels out the other "QC"
(qc.yaml).

Signed-off-by: Paul Spooren <mail@aparcar.org>

ilka-schulz commented:
closed due to inactivity
